packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,269 @@
+import path from "node:path";
+import fg from "fast-glob";
+
+import type { ConfigRoute, RouteManifest } from "./routes";
+import {
+  escapeEnd,
+  escapeStart,
+  isSegmentSeparator,
+  optionalEnd,
+  optionalStart,
+  paramPrefixChar,
+  routeModuleExts,
+} from "./routesConvention";
+
+export function flatRoutes(
+  appDirectory: string,
+  ignoredFilePatterns?: string[]
+): RouteManifest {
+  let extensions = routeModuleExts.join(",");
+
+  let routePaths = fg.sync(`*{${extensions}}`, {
+    absolute: true,
+    cwd: path.join(appDirectory, "routes"),
+    ignore: ignoredFilePatterns,
+  });
+
+  return flatRoutesUniversal(appDirectory, routePaths);
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory - The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routes: Record<string, ConfigRoute> = {};
+  let sortedRoutes = routePaths
+    .sort((a, b) => (a.length - b.length > 0 ? 1 : -1))
+    .map((routePath) =>
+      routePath.slice(appDirectory.length + 1).replace(/\\/g, "/")
+    );
+  let processedIds: string[] = [];
+  for (let routePath of sortedRoutes) {
+    let routeId = routeIdFromPath(routePath);
+    if (routes[routeId]) {
+      throw new Error(`Duplicate route: ${routeId}`);
+    }
+
+    let tmpId = "";
+    let parentId = "";
+    for (let processedId of processedIds) {
+      if (routeId.startsWith(processedId.replace(/\$$/, "*"))) {
+        let [segments, delimiters] = splitByDelimiter(routeId);
+        if (segments.at(1)?.endsWith("_")) {
```

Comment:
What if the `.` is escaped? E.g. `app.hello_[.]world`? Then this check would get incorrectly triggered, no?

Reply:
Also, don't we have to account for ending `_` for other segments? not just the second one (`at(1)`)?